flik/optimize/functions.py

```diff
@@ -0,0 +1,512 @@
+#
+# Flik is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 3
+# of the License, or (at your option) any later version.
+#
+# Flik is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, see <http://www.gnu.org/licenses/>
+
+
+r"""Optimization functions."""
+
+
+from numbers import Integral
+from numbers import Real
+
+import numpy as np
+
+from flik.objective import Objective
+from flik.objective import ForwardDiffObjective
+from flik.objective import CentralDiffObjective
+
+from flik.jacobian import Jacobian
+from flik.jacobian import CentralDiffJacobian
+
+from flik.gradient import Gradient
+from flik.gradient import ForwardDiffGradient
+from flik.gradient import CentralDiffGradient
+
+from flik.hessian import Hessian
+from flik.hessian import ForwardDiffHessian
+from flik.hessian import CentralDiffHessian
+
+from flik.methods import Method
+from flik.methods import BroydenFamilyMethod
+from flik.methods import BofillMethod
+from flik.methods import HuangMethod
+from flik.methods import LBfgsMethod
+from flik.methods import KrylovMethod
+from flik.methods import DiisMethod
+
+from flik.linesearch import LineSearch
+from flik.linesearch import ConstantLineSearch
+from flik.linesearch import BacktraceLineSearch
+
+from flik.trustregion import TrustRegion
+
+from flik.optimize.exceptions import LinAlgError
+
+
+__all__ = [
+    "root",
+    "minimize",
+    ]
+
+
+# Message for reaching maximum number of iterations
+MSG_MAXITER_REACHED = "Maximum number of iterations reached."
+
+
+# Message for failure to converge the system
+# TODO: Change to something more general (and get the tests to use these
+# MSG variable definitions, i.e., from flik.optimize.functions import MSG_...
+MSG_NO_CONVERGE = "Singular Jacobian; no solution found."
+
+
+# Message for a converged system
+MSG_CONVERGED = "Convergence obtained."
+
+
+def root(f, x, J=None, cost=None, tol=1.0e-9, maxiter=100, method=None,
+         linesearch=None, trustregion=None, lsopts=None, tropts=None,
+         **kwargs):
+    r"""
+    Solve a system of nonlinear equations.
+
+    Parameters
+    ----------
+    f : callable
+        Vector-valued function corresponding to system of equations.
+    x : (np.ndarray | int)
+        Solution initial guess vector or solution vector size.
+    J : (callable | Jacobian | str), optional
+        Jacobian of function ``f``, ``J(x)`` if callable or Jacobian instance.
+        Finite difference method ("forward" | "central") if str.
+        By default, a central difference Jacobian will be used where necessary.
+    cost : callable, optional
```

Comment:
For example, `cost` might be the energy function, if you have some nonlinear system of equations corresponding to a (variational) wavefunction optimization.